### PR TITLE
#168781708-user can book an accommodation

### PR DESCRIPTION
**Pull request by sabin18**
#### What does this PR do?
- It allow user to view all available accommodations and book an accommodation.
#### Description of Task to be completed?
- add view all available accommodation function
- add booking function
- add view all booking made by the user function
#### How should this be manually tested?
- clone the repository : git clone https://github.com/andela/caret-bn-backend.git
- checkout to branch ft-book-accommodation-168781708
- run sequelize db:migrate
- run sequelize db:seed:all
- Login 
- Take the token to your headers in authorization as Bearer Token to each routes of this feature
- use this route to view all available accommodations `GET: /api/v1/accommodations/available`
- use this route to view all your bookings  `GET: /api/v1/accommodations/bookings`
- use this route to book an accommodation `PATCH:/api/v1/accommodations/book`
- on booking an accommodation you add the Id of accommodation in  the body.   
#### Any background context you want to provide?
`N/A`
#### What are the relevant pivotal tracker stories?
[168781708](https://www.pivotaltracker.com/story/show/168781708)
#### Screenshots (if appropriate)
Screen Shot 2019-10-28 at 10 34 44
Screen Shot 2019-10-28 at 10 35 11
Screen Shot 2019-10-28 at 10 35 32

#### Questions:
`N/A`

### CODE REVIEW

**Review by ngireric123 — APPROVED** [review]

File: src/controllers/accommodationController.js [hunk 1]
```diff
@@ -107,4 +110,70 @@ export default class AccommodationController {
...
+      },
+      attributes: { exclude: ['locationId', 'owner'] },
+      include: [{
+        model: models.locations, as: 'accommodationLocation', attributes: ['id', 'name']
+      }]
+    }).then(accommodation => responseUtil(res, 200, FOUND, accommodation));
+  }
+
+  // book an accomodation
+  static BookAccommdation(req, res) {
```
Comment (ngireric123) [inline src/controllers/accommodationController.js]: follow naming convention we use

**Review by alainmateso — CHANGES_REQUESTED** [review]
Good job @sabin18 . However, please ensure that you make the necessary validations. For example, I can book an accommodation with dates that have already passed. Please look into that. Also, the error messages returned on the wrong date formats should be a date format not a date itself.

**Review by alainmateso — APPROVED** [review]

File: src/controllers/accommodationController.js [hunk 2]
```diff
@@ -107,4 +110,73 @@ export default class AccommodationController {
...
+  // view available accommodations
+  static availableAccommdation(req, res) {
+    const { Op } = sequelize;
+    models.accommodations.findAll({
+      where: {
+        availableSpace: {
+          [Op.gt]: 0,
+        }
+      },
+      attributes: { exclude: ['locationId', 'owner'] },
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js:125]: Add the owner object to the response

File: src/controllers/accommodationController.js [hunk 3]
```diff
@@ -107,4 +110,73 @@ export default class AccommodationController {
...
+        model: models.locations, as: 'accommodationLocation', attributes: ['id', 'name']
+      }]
+    }).then(accommodation => responseUtil(res, 200, FOUND, accommodation));
+  }
+
+  // book an accomodation
+  static bookAccommdation(req, res) {
+    const { Op } = sequelize;
+    const { id } = req.params;
+    const { checkInDate, checkOutDate } = req.body;
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: Add `accomodationId` and `numberRooms` to the req.body

File: src/controllers/accommodationController.js [hunk 4]
```diff
@@ -107,4 +110,73 @@ export default class AccommodationController {
...
+      }).then(accommodation => {
+        if (accommodation.length === 0) {
+          return responseUtil(res, 400, strings.accommodation.error.NOT_AVAILABLE);
+        }
+        if (moment(checkOutDate) < moment.now() || moment(checkOutDate) < moment.now()) {
+          return responseUtil(res, 400, strings.accommodation.error.INVALID_DATE);
+        }
+        if (moment(checkOutDate) < moment(checkInDate)) {
+          return responseUtil(res, 400, strings.accommodation.error.DATE_ERROR);
+        }
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: Make this a helper function to be reused

File: src/controllers/accommodationController.js [hunk 5]
```diff
@@ -107,4 +110,73 @@ export default class AccommodationController {
...
+          id,
+          availableSpace: {
+            [Op.gt]: 0,
+          }
+        }
+      }).then(accommodation => {
+        if (accommodation.length === 0) {
+          return responseUtil(res, 400, strings.accommodation.error.NOT_AVAILABLE);
+        }
+        if (moment(checkOutDate) < moment.now() || moment(checkOutDate) < moment.now()) {
```
Comment (nakiwuge) [inline src/controllers/accommodationController.js]: Make this a helper function to be reused

File: src/routes/api/accommodations.js [hunk 6]
```diff
@@ -16,11 +16,14 @@ const {
   createAccommodation,
   getAllAccommodations,
   editAccommodation,
-  deleteAccommodation
+  deleteAccommodation,
+  availableAccommdation,
+  bookAccommdation
+
```
Comment (nakiwuge) [inline src/routes/api/accommodations.js]: remove this line

File: src/routes/api/accommodations.js [hunk 7]
```diff
@@ -190,5 +193,7 @@ router.post('/', validateToken, checkSupplierRole, catchEmptyForm, multipartMidd
 router.get('/', validateToken, getAllAccommodations);
 router.patch('/:id/edit', validateToken, checkId, catchEmptyForm, multipartMiddleware, isAccommodationFound, isOwner, validateAccommodationEdit, editAccommodation);
 router.delete('/:id/delete', validateToken, checkId, isAccommodationFound, isOwner, deleteAccommodation);
+router.get('/available', validateToken, availableAccommdation);
+router.patch('/booking/:id', validateToken, checkId, validateDate, bookAccommdation);
```
Comment (nakiwuge) [inline src/routes/api/accommodations.js]: Change this route to `/book` and remove the param

**Review by nakiwuge — CHANGES_REQUESTED** [review]
Good job @sabin18, I have left comments on this pr. Also kindly add a route `/bookings` that will allow users to view their bookings

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**